weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -1978,6 +1911,329 @@ def completions_create(
+                if function_data.get("name"):
+                    existing_tool_call["function"]["name"] = function_data["name"]
+                if "arguments" in function_data:
+                    existing_tool_call["function"]["arguments"] += function_data[
+                        "arguments"
+                    ]
+
+    def _clean_tool_calls(
+        self, tool_calls: list[dict[str, Any]]
+    ) -> list[dict[str, Any]]:
+        """Clean up tool_calls - remove incomplete ones and ensure proper format."""
+        cleaned_tool_calls = []
+        for tool_call in tool_calls:
+            if tool_call.get("id") and tool_call.get("function", {}).get("name"):
+                cleaned_tool_call = {
+                    "function": {
+                        "arguments": tool_call["function"]["arguments"],
+                        "name": tool_call["function"]["name"],
+                    },
+                    "id": tool_call["id"],
+                    "type": "function",
+                }
+                cleaned_tool_calls.append(cleaned_tool_call)
+        return cleaned_tool_calls
+
+    def _build_aggregated_output(
+        self,
+        aggregated_metadata: dict[str, Any],
+        assistant_acc: list[str],
+        tool_calls: list[dict[str, Any]],
+        chunk: dict[str, Any],
+    ) -> dict[str, Any]:
+        """Build the aggregated output from accumulated data."""
+        current_finish_reason = chunk.get("choices", [{}])[0].get("finish_reason")
+        cleaned_tool_calls = self._clean_tool_calls(tool_calls)
+
+        return {
+            "id": aggregated_metadata.get("id", ""),
+            "created": aggregated_metadata.get("created", 0),
+            "model": aggregated_metadata.get("model", ""),
+            "object": "chat.completion",
+            "system_fingerprint": aggregated_metadata.get("system_fingerprint", ""),
+            "choices": [
+                {
+                    "finish_reason": current_finish_reason,
+                    "index": 0,
+                    "message": {
+                        "content": (
+                            "".join(assistant_acc)
+                            if assistant_acc
+                            else (None if cleaned_tool_calls else "")
+                        ),
+                        "role": "assistant",
+                        "tool_calls": (
+                            cleaned_tool_calls if cleaned_tool_calls else None
+                        ),
+                        "function_call": None,
+                    },
+                }
+            ],
+            "usage": aggregated_metadata.get("usage", {}),
+            "service_tier": aggregated_metadata.get("service_tier", "default"),
+        }
+
+    def _create_tracked_stream_wrapper(
+        self,
+        chunk_iter: Iterator[dict[str, Any]],
+        start_call: CallStartCHInsertable,
+        model_name: str,
+        project_id: str,
+    ) -> Iterator[dict[str, Any]]:
```

Comment:
I would argue this should also be a standalone function. Just pass `self._insert_call` has an input.